src/NServiceBus.Core/Persistence/InMemory/Gateway/ClientIdStorage.cs

```diff
@@ -0,0 +1,58 @@
+﻿namespace NServiceBus
+{
+    using System.Collections.Generic;
+
+    class ClientIdStorage
+    {
+        public ClientIdStorage(int maxSize)
+        {
+            this.maxSize = maxSize;
+        }
+
+        public bool IsDuplicate(string clientId)
+        {
+            lock (clientIdSet)
+            {
+                // This implementation is not 100% since there is a race condition here where another thread might consider the same message as a non-duplicate.
```

Comment:
```suggestion
                // This implementation is not 100% consistent since there is a race condition here where another thread might consider the same message as a non-duplicate.
```